### ISSUE TITLE
PDF.dumps saves PDF with pages out of order

### ISSUE THREAD

**Issue by scbarton** [jorisschellekens/ptext-release#3]
Using a 20-page PDF, loading and saving the file results in pages out of order:

```
from ptext.pdf.pdf import PDF

doc=None
with open(pdf_file, "rb") as in_file_handle:
    doc = PDF.loads(in_file_handle)

with open("output.pdf", "wb") as out_file_handle:
        PDF.dumps(out_file_handle, doc)
```

Is there any way to control this?

**Comment by jorisschellekens** [comment]
Normally, a PDF document defines a Pages dictionary, containing an array of Page objects.
However, the PDF specification also allows Page objects to be nested. In my experience (I have a repo containing 2K documents I gathered "in the wild"), this is not the most common use-case.

One of the transformations pText does when reading a PDF is attempt to flatten these tree-like Page structures.
I guess there must be a bug in that mechanism.

If you yourself want to have a look, it would be in `ReadRootDictionaryTransformer` starting at line 75.

If you prefer I have a look at it, it would be convenient to have an example document where it goes wrong.

Kind regards,
Joris

**Comment by scbarton** [comment]
Thanks, I'll take a look at that function, though I don't have much expertise.  The PDF I have been using as an example is [this one](https://issuu.com/ecs1902/docs/2021-orlando-cfp-03).

**Comment by jorisschellekens** [comment]
I'll have a look at it.
But keep in mind that pText is not my day-job, so I can look into it only in the evenings or weekends.
Thanks for sharing the PDF, I'll add it to my dataset.

**Comment by jorisschellekens** [comment]
I fixed the issue.
There was a small mistake in how the aforementioned code handled depth first search.

It used to do this:

    stack_to_handle.extend(transformed_root_dictionary["Pages"]["Kids"])

    # DFS
    while len(stack_to_handle) > 0:
        obj = stack_to_handle.pop(0)
        if isinstance(obj, Page):
            pages_in_order.append(obj)
        # \Pages
        if (
            isinstance(obj, Dictionary)
            and "Type" in obj
            and obj["Type"] == "Pages"
            and "Kids" in obj
            and isinstance(obj["Kids"], List)
        ):
            for k in obj["Kids"]:
                stack_to_handle.insert(0, k)

and now it does this:

    stack_to_handle.append(transformed_root_dictionary["Pages"])

    # DFS
    while len(stack_to_handle) > 0:
        obj = stack_to_handle.pop(0)
        if isinstance(obj, Page):
            pages_in_order.append(obj)
        # \Pages
        if (
            isinstance(obj, Dictionary)
            and "Type" in obj
            and obj["Type"] == "Pages"
            and "Kids" in obj
            and isinstance(obj["Kids"], List)
        ):
            for k in obj["Kids"]:
                stack_to_handle.append(k)

I verified it on your example document, and it seems to work perfectly now.
The new version is available on PyPi as 1.8.4.beta.

I still need to sync the private repo with the public one.

**Comment by scbarton** [comment]
Yes, I can confirm that it is fixed, thanks!  

However, I did get a bunch of messages during `PDF.dumps`:
```
'decimal.Decimal' object has no attribute 'set_reference'
'Boolean' object has no attribute 'set_reference'
'bytes' object has no attribute 'set_reference'
```

The output looks fine but I wonder if something may be broken at a lower level.
